File: This&that/_scripting_languages/Ruby/design_patterns/visitor.rb
#
# Visitor Patterns
#     - have and object1 that accept object2, object2 will change the sate of object1
#


#
# Class that will be visited
#
class Car 
    attr_accessor :gear

    def initialize
        @gear = nil
    end

    #
    # The visitor will enter this object and make changes to it.
    #
    def accept(visitor)
        visitor.visit(self)
    end

    def change_gear(gear)
        @gear = gear
    end
end

#
# Visitor
#
class Person 
    attr_accessor :name

    def initialize(name)
        @name = name
    end

    def visit(car)
        puts "#{@name} is changing gears in the car"
        car.change_gear(1)
    end
end

car = Car.new
brian = Person.new('Brian')
car.accept(brian) # Brian is changing gears in the car

File: This&that/_scripting_languages/Javascript/_design_patterns/visitor.js
// I STILL NEED TO TEST THIS OUT......


// http://www.dofactory.com/javascript/design-patterns


// have a person object, that has a accept(visitor) method
// have a visitor object that has a visit(person) method

// a visitor is just like an external helper...

// the story is:  person, have a habit of acceptiong visitors
//    and the visitors will change stuff around your life!

// main object
function Person(name, num){
    this.num = num;
    var self = this;

    this.getNum = function(){
        return this.num;
    }

    this.setNum = function(num){
        this.num = num;
    }

    this.accept = function(visitor){
         visitor.visit(self);
    }
}

// visitor
function Visitor(){
    this.visit = function(person){
        person.setNum(person.getNum() + 1000);
    }
}

// run
var brian = new Person('brian', 10);
var visitor = new Visitor();
brian.accept(visitor);
brian.getNum();


